mod.go

```diff
@@ -0,0 +1,579 @@
+// multithreaded settings.
+//
+// Adjust cacheIndexBits and cacheWays to scale the size of the cache.
+// Total cache size is (24+72*cacheWays)*2^cacheIndexBits bytes, which is
+// 96 KiB with cacheIndexBits = 8 and cacheWays = 5. There are also 16 bytes
+// for hit and miss counters.
+//
+// Reasonable values are quite small, e.g. cacheIndexBits from 2 to 10, and
+// cacheWays around 5. Note that cacheWays = 5+8n makes the set size an integer
+// number of 64-byte cachelines.
+
+const (
+	cacheIndexBits = 8
+	cacheWays      = 5
+
+	cacheSets = 1 << cacheIndexBits
+	cacheMask = cacheSets - 1
+)
+
+type cacheSet struct {
+	rw  sync.RWMutex
+	mod [cacheWays]Int
+	inv [cacheWays][5]uint64
+}
+
+type reciprocalCache struct {
+	set [cacheSets]cacheSet
+	hit  uint64
+	miss uint64
+}
+
+func (c *reciprocalCache) Stats() (hit, miss uint64) {
+	return c.hit, c.miss
+}
+
+func (cache *reciprocalCache) has(m Int, index uint64, dest *[5]uint64) bool {
+	cache.set[index].rw.RLock()
+	defer cache.set[index].rw.RUnlock()
+
+	for w := 0; w < cacheWays; w++ {
+		if cache.set[index].mod[w].Eq(&m) {
+			copy(dest[:], cache.set[index].inv[w][:])
+			cache.hit++
+			return true
+		}
+	}
+	cache.miss++
+	return false
+}
+
+func (cache *reciprocalCache) put(m Int, index uint64, mu [5]uint64) {
+	cache.set[index].rw.Lock()
+	defer cache.set[index].rw.Unlock()
+
+	var w int
+	for w = 0; w < cacheWays; w++ {
+		if cache.set[index].mod[w].IsZero() {
+			// Found an empty slot
+			cache.set[index].mod[w] = m
+			cache.set[index].inv[w] = mu
+			return
+		}
+	}
+	// Shift old elements, evicting the oldest
+	for w = cacheWays - 1; w > 0; w-- {
+		cache.set[index].mod[w] = cache.set[index].mod[w-1]
+		cache.set[index].inv[w] = cache.set[index].inv[w-1]
+	}
+	cache.set[index].mod[0] = m
+	cache.set[index].inv[0] = mu
+}
+
+var cache reciprocalCache
+
+// Some utility functions
+
+func leadingZeros(x Int) (z int) {
+	z  = bits.LeadingZeros64(x[3]); if z <  64 { return z }
+	z += bits.LeadingZeros64(x[2]); if z < 128 { return z }
+	z += bits.LeadingZeros64(x[1]); if z < 192 { return z }
+	z += bits.LeadingZeros64(x[0]); return z
+}
+
+func onesCount(x Int) (z int) {
+	z =	bits.OnesCount64(x[0]) +
+		bits.OnesCount64(x[1]) +
+		bits.OnesCount64(x[2]) +
+		bits.OnesCount64(x[3])
+	return z
+}
+
+// shiftright320 shifts a 320-bit value 0-63 bits right
+// z = x >> (s % 64)
+
+func shiftright320(x [5]uint64, s uint) (z [5]uint64) {
+	r := s % 64	// right shift
+	l := 64 - r	// left shift
+
+	z[0] = (x[0] >> r) | (x[1] << l)
+	z[1] = (x[1] >> r) | (x[2] << l)
+	z[2] = (x[2] >> r) | (x[3] << l)
+	z[3] = (x[3] >> r) | (x[4] << l)
+	z[4] = (x[4] >> r)
+
+	return z
+}
+
+// reciprocal computes a 320-bit value representing 1/m
+//
+// Notes:
+// - reciprocal(0) = 0
+// - reciprocal(1) = 2^320-1
+// - otherwise, the result is normalised to have non-zero most significant word
+// - starts with a 32-bit division, refines with newton-raphson iterations
```

Comment:
This computes `2^512/m` right? The Handbook of Applied Cryptography does not prescribe how to compute it. Why this methods has been chosen?

Reply:
> This computes `2^512/m` right? The Handbook of Applied Cryptography does not prescribe how to compute it. Why this methods has been chosen?

Yes. The basic idea is to start with a rough approximation and then repeatedly refine it with Newton-Raphson (see https://en.wikipedia.org/wiki/Division_algorithm#Newton%E2%80%93Raphson_division and http://www.quadibloc.com/comp/cp0202.htm). I looked at some other alternatives too before landing on this algorithm, since it is fast and does not involve any complicated bit manipulation. It is also used in hardware, starting with a small lookup table. The adaptations to this context are my own.